### DISCUSSION TITLE
Neoscore 0.1.7 released with improved rotation and scaling support!

### DISCUSSION THREAD

**Discussion by ajyoon** [DigiScore/neoscore#73] [Announcements]
I've just release 0.1.7, which includes a major improvement to the way neoscore handles rotation and scaling. Previously, there was no way to rotate or scale a group of objects together. An object rotated would not take its children with it, and similarly with scaling. In the new release we support this behavior outside of flowable contexts.

![rotation](https://user-images.githubusercontent.com/15976214/193721305-3cdf9905-1518-4ed1-9dca-4023e6d9b808.png)

That said, this functionality remains fairly experimental. There are many things that will not work as expected with this, like paths with elements parented to other objects. This is also a complicated change, so there's a good chance there are bugs - please report anything that seems wrong!

[You can read more about this new behavior at the docs page here](https://neoscore.org/core_concepts/rotation_and_scaling.html).

This release also includes a few other improvements:
- [Thick pens now display correctly](https://github.com/DigiScore/neoscore/issues/14) thanks to @Xavman42 
- Paths now support scaling as well
- A memory leak in flowable layout has been fixed
- Scaling and rotation can now be performed relative to a given `transform_origin` point thanks to @Xavman42 

## ⚠️ Breaking changes ⚠️

- For complicated internal reasons, explicit `z_index` support has been removed. Item rendering order is now determined exclusively by the order in which objects appear in the document tree's depth-first traversal. Within each page, objects created first, and their children (created first) will be rendered first.
- This release substantially reworks the way interface-layer classes work, as they now sometimes have parents. The argument order for many interface class init functions has been changed.
- This release also significantly changes the way `PositionedObject.render_complete()` works. **Classes which implement this method will need to be updated**. See its documentation for more.

**Comment by ajyoon** [comment]
@Xavman42 your project got me thinking that with this new functionality it would be pretty easy to build fractal scores using staves in place of line segments. I threw together [a little example script for this](https://github.com/DigiScore/neoscore/blob/main/examples/fractal_score.py) you may be interested in. Lots of very interesting ideas in this area!

![fractal_score](https://user-images.githubusercontent.com/15976214/194176100-b6f82015-1882-435a-b4e9-5c01f1e385c6.png)

@craigvear

**Reply by Xavman42** [reply]
This is phenomenal! It made me realize that there are probably some Crumb scores that could be replicated now...


![image](https://user-images.githubusercontent.com/79334532/194200493-62798470-62e2-45c5-b801-310ae98d504b.png)

**Reply by ajyoon** [reply]
Definitely! now doing Crumb-style curved staves would be a whole other challenge, but probably achievable...

**Comment by craigvear** [comment]
Wow. This looks great guys. Very exciting. I like the fractal Andrew. Might develop it further as a taster score for the forth coming series of DigiScore workshops 🥳

**Comment by Xavman42** [comment]
Thought of a nice-to-have feature:

Rather than initialize objects then change their rotation later, it would be nice to be able to set rotation in the initial class call. For example, this:
```
s = Staff(Point(Mm(x), -Mm(y)), None, Mm(length))
s.rotation = -angle
```
could be achieved with this:
```
Staff(Point(Mm(x), -Mm(y)), None, Mm(length), -angle))
```
or something similar.

**Reply by ajyoon** [reply]
This was somewhat of an intentional design choice since complex objects like staves have so many properties the signatures would become very verbose if we included everything there. In general, for low level (core module) objects, we try to include everything in the init signature, but higher level objects we prefer to only include higher level fields in the init. If enough people request it though we can reconsider.
